### PR TITLE
Add a new line at the end of error reports (#6823

### PR DESCRIPTION
**Pull request by esope**
This is an attempt to fix issue #6158

Remark: If a command (say, `ocamlopt`) returns several messages (such as warnings), a single line is put at the end, since this is considered as a single message.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Thanks for trying to address this. All the gratuitous newlines are a bit annoying. Can we print the `cut` only as a separator between errors?

@snowleopard what do you think of this change in principle? We can make it console backend dependent to avoid any inconvenience it might cause inside at jst for you.

**Comment by esope** [comment]
I agree too that some of the added newlines are annoying, especially terminal ones.
I have not found how to implement the "add new lines only between messages" behavior: it seems there is no easy way to detect which message is the "last" one.
A different approach would be to add a new line _before_ a message. Maybe this would feel less annoying.

**Comment by rgrinberg** [comment]
The behavior you describe sounds reasonable. In other words "Prepend a newline if we aren't the first message being printed".

**Comment by esope** [comment]
OK, I will try with this approach when I find some free time, then.

**Comment by esope** [comment]
This seems to bring much more reasonable results.
I am not sure that the use of a reference in a function that is passed to `Fiber.Mutex.with_lock` is correct. What is the best practice that you use in the dune codebase?
I will add a test as soon as possible.

**Comment by rgrinberg** [comment]
It's correct. As long as it doesn't cross a bind, there's nothing to worry about.

However, why did you choose to put this change in build_system.ml? I would expect `Dune_util.Report_error.report` to be the place to do it.

**Comment by esope** [comment]
OK, I moved the change to `Dune_util.Report_error.report`.

File: src/dune_util/report_error.ml
Comment (snowleopard) [inline src/dune_util/report_error.ml:150]: The reference never gets reset back to `true`, so I guess this feature doesn't interact well with file-watching builds. You'll need to add and use a mechanism for correctly handling rebuilds.

**Review by snowleopard — CHANGES_REQUESTED** [review]
I think we should either support file-watching builds, or disable this feature during file-watching builds, because the current implementation seems broken.

**Comment by rgrinberg** [comment]
@snowleopard sounds good, but do you think we need a flag for this feature? Or are you happy changing the output internally as well.

**Comment by snowleopard** [comment]
> @snowleopard sounds good, but do you think we need a flag for this feature? Or are you happy changing the output internally as well.

Yes, it seems nice to keep this feature behind a flag, at least for some time, since we are technically breaking Dune's public interface. For example, various tests calling out to Dune might break.

Internally, I think there is a high chance we would want to enable this feature.

**Comment by esope** [comment]
Thanks for the feedback.
What should be the name of this flag? `--separate-error-messages`? Should this be controlled by an option in `dune-workspace` too?
In watch mode only, when this flag is on, what do you think about _always_ prepending the error message with a blank line?

**Comment by rgrinberg** [comment]
1. That name sounds fine.
2. For now I wouldn't worry about this and we can even just add an environment variable to control this. If this setting is good in practice, we're just going to make it the default.
3. Why does this extra newline help in watch mode?

**Comment by esope** [comment]
Thanks.
About point 3, this would remove the need of keeping an internal state (a `bool ref` in this case) when in watch mode. This means that there would be mo problem anymore related to memoization or rebuilds. Or am I mistaken?

**Comment by esope** [comment]
OK, I added the flag `--separate-error-messages` to control the new behavior of `Dune_util.Report.report`.
The flag is always off when watch mode is enabled.

File: bin/common.ml
Comment (snowleopard) [inline bin/common.ml:816]:
```suggestion
               currently disabled in the file-watching mode.")
```

File: test/blackbox-tests/test-cases/error_messages_separated.t
Comment (snowleopard) [inline test/blackbox-tests/test-cases/error_messages_separated.t:70]:
```suggestion
the first message, and no blank line is inserted after the last
```

File: bin/common.ml
Comment (snowleopard) [inline bin/common.ml:853]:
```suggestion
        (* This flag is currently disabled in the file-watching mode. To be fixed. *)
```

**Review by snowleopard — COMMENTED** [review]
Thanks, looks good now! I left a couple of suggestions.

It shouldn't be hard to make this flag work for the file-watching mode, I think. Perhaps, `Memo.Run.t` can be stored alongside the `bool ref` to decide when that `ref` gets stale and needs to be reset.

**Comment by esope** [comment]
Thanks for the feedback, @snowleopard . I implemented your suggested fixes.

**Review by snowleopard — COMMENTED** [review]
Thanks! I'll leave it to @rgrinberg to approve and merge. (As far as I'm concerned there are no blockers any more.)

**Comment by esope** [comment]
@rgrinberg, I'm happy to integrate more changes, should you have any.

**Comment by rgrinberg** [comment]
Tbh, I think we need to make sure this feature works with watch mode before we consider it for merging. Watch mode is the recommended way to use dune, and whatever feature we're adding should work with it out of the box.

**Comment by esope** [comment]
Understood. I'll have a second look at it, then.

**Comment by Alizter** [comment]
I was also going to mention that it might be a good idea to allow for a config option, so that users can enable this by default if they desire.

**Comment by esope** [comment]
OK, I've pushed some modification to support file-watching mode.
When file-watching mode is disabled, the behavior is the same as before (a blank line is inserted between error messages).
When file-watching mode is enabled, then a blank line is always printed _after_ each error message. This makes the implementation correct wrt caching, and I think this gives a consistent experience to users.

I'll wait for your feedback before dealing with config option issue.

File: src/dune_util/report_error.ml
Comment (Alizter) [inline src/dune_util/report_error.ml:138]: Make the comments in this file `(**`

**Review by Alizter — APPROVED** [review]

File: src/dune_util/report_error.ml
Comment (esope) [reply src/dune_util/report_error.ml]: Done in the last commit.

**Comment by rgrinberg** [comment]
@gasche what do you think about the behavior proposed for watch mode? It seems alright to me.

**Comment by gasche** [comment]
Thoughts (but: remember that I know very little about dune internals):

1. The observed behavior looks nice to me, and the compromise of having a different behavior when watching or not seems reasonable.
2. I wonder how the compiler should deal with the same feature requirement. In particular I wondered if there was a solution in the compiler that is satisfying and would require no change to dune. (Then we should do it this way rather than changing dune.) But unfortunately I don't see it. We could take inspiration of the logic here and print a line before each message that is not the first, and I *think* that the two logics would compose nicely. (Or we could add a blank line after each message, and then dune should not do anything, but we have the same problem as in @esope's first attempt that it's not so nice.)
3. Looking at the implementation, I wonder why the change is at the "report error" level rather than the `Console.print_user_message` level. It looks like the Console level (1) sees more messages that may also benefit from newline separation, and (2) may have precise knowledge of when the console is cleared, and the ability to accurately maintain state on whether we are the first message or not.

In any case, thanks everyone for looking at this. It's a small thing, but I think it could make a positive difference especially for beginner OCaml users.

**Comment by emillon** [comment]
@esope do you think you'll be able to fix this this week? otherwise I'm happy to move this back to 3.8 (at the moment the 3.7 release is blocked on this feature). Thanks!

**Comment by esope** [comment]
I can find some time next week. Sorry I've been pretty busy and unresponsive lately.
It is ok for me to leave this to 3.8.

**Comment by esope** [comment]
I moved the logic for adding blank lines to `Dune_console.print_user_message` as suggested by @gasche .
This should cover more cases than before. I see no difference in practice with the previous version.

I have not tried to have different behaviors for the different console backends, because I don't see what could be specifically done for a particular backend, and because this would necessitate to duplicate code into the different backends, I think.
You guys might have different opinions, though.

**Comment by gasche** [comment]
Two comments from an outsider:
- Nitpick: the name `error` is less appropriate at this abstraction level, maybe `separate_error_messages => separate_messages` and `first_error` => `first_message`?
- Logic: maybe you could set `first_message := true` in the `reset` and `reset_flush_history` functions, and this code would magically work for the watch mode as well? (I'm assuming the watch mode is implemented by calling `reset` to clear the terminal). Then you could simplify your code quite a bit.

**Comment by snowleopard** [comment]
I quite like both of @gasche's suggestions!

**Comment by esope** [comment]
Very good remarks, indeed! Thanks!
Both are now implemented.

**Comment by esope** [comment]
I can hear no more requests for changes, does this mean everybody is happy about this implementation?
If this is a yes, this should probably be merged.

I will be happy to add an environment variable or an option in `dune-project` in a subsequent PR to control this setting more easily, if you think this is useful.

**Comment by esope** [comment]
@snowleopard I updated the (now quite old) changes you requested, as well as @gasche 's.
I would very much like, that we make progress on this issue.

File: CHANGES.md
Comment (snowleopard) [inline CHANGES.md:11]:
```suggestion
- Allow running `$ dune exec` in watch mode (with the `-w` flag). In watch mode,
```

File: src/dune_console/dune_console.ml
Comment (snowleopard) [inline src/dune_console/dune_console.ml:227]:
```suggestion
  (* forget that [print_user_message] has ever been called *)
  first_msg := true;
```

File: src/dune_console/dune_console.ml
Comment (snowleopard) [inline src/dune_console/dune_console.ml:221]:
```suggestion
  (* forget that [print_user_message] has ever been called *)
  first_msg := true;
```

**Review by snowleopard — APPROVED** [review]
Thanks! It's great that this feature is now compatible with the file-watching mode.

I wonder if `--separate-error-messages` is still the right name for the flag though. It now affects all messages, not just error messages, so perhaps it should be called `--separate-messages`? I don't have a strong opinion about this.

**Comment by rgrinberg** [comment]
Thanks for taking a look. I'll rebase and merge.

> It now affects all messages, not just error messages, so perhaps it should be called --separate-messages? I don't have a strong opinion about this.

I'll call it `--display-separate-messages`.  Without the "display', it's not clear which part of dune this is affecting exactly.

**Comment by esope** [comment]
Thanks. I will rename the flag as soon as possible.

**Comment by rgrinberg** [comment]
I already did the rename

**Comment by esope** [comment]
Great! Thanks!
I am eager to hear the feedback from users of JaneStreet and Tarides.
@gasche : should I/you try to implement the "blank line separation" on the OCaml compiler side?

**Comment by gasche** [comment]
I proposed something like this yesterday in https://github.com/ocaml/ocaml/pull/12024

This is a bit harder to do on the compiler side because we don't have a generic abstraction of "message", we have the "reports" (errors and warnings) that are easy to handle, and a ton of other ad-hoc printing calls that would each need proper logic to be handled as messages. The PR above is a first iteration that does a good enough job, I hope.

**Comment by gasche** [comment]
The compiler PR has been merged thanks to a careful review of @Octachron. Blank lines for everyone!

**Comment by snowleopard** [comment]
Amazing, thanks @gasche! Great to see such tight collaboration between compiler and build system devs.

**Pull request merged**
